File: dfs.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def dfs(root, target):
    if root is None:
        return

    if root.value == target:
        return True
    dfs(root.left, target)
    dfs(root.right, target)


def dfs_iter(root):
    if root is None:
        return

    stack = [root]

    while stack:
        node = stack.pop()
        print(node.value)

        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)

File: dsa/dfs.ts
type DFSTreeNode<T> = {
  val: T;
  left: DFSTreeNode<T> | null;
  right: DFSTreeNode<T> | null;
}

function dfsPreorder<T>(root: DFSTreeNode<T> | null): T[] {
  if (!root) return [];

  const result: T[] = [];
  const stack: Array<DFSTreeNode<T>> = [root];

  while (stack.length) {
    const node = stack.pop();
    if (!node) break;

    result.push(node.val); 
    if (node.right) stack.push(node.right);
    if (node.left) stack.push(node.left);
  }

  return result;
}

function preorder<T>(root: DFSTreeNode<T> | null, out: T[] = []): T[] {
  if (!root) return out;
  out.push(root.val);
  preorder(root.left, out);
  preorder(root.right, out);
  return out;
}